win32ss/gdi/eng/device.c

```diff
@@ -32,6 +32,133 @@ InitDeviceImpl(VOID)
     return STATUS_SUCCESS;
 }
 
+static
+BOOLEAN
+EngpHasVgaDriver(
+    IN PGRAPHICS_DEVICE pGraphicsDevice)
+{
+    WCHAR awcDeviceKey[256], awcServiceName[100];
+    PWSTR lastBkSlash;
+    NTSTATUS Status;
+    ULONG cbValue;
+    HKEY hkey;
+
+    /* Open the key for the adapters */
+    Status = RegOpenKey(L"\\Registry\\Machine\\HARDWARE\\DEVICEMAP\\VIDEO", &hkey);
+    if (!NT_SUCCESS(Status))
+    {
+        ERR("Could not open HARDWARE\\DEVICEMAP\\VIDEO registry key: 0x%08lx\n", Status);
+        return FALSE;
+    }
+
+    /* Read the name of the device key */
+    cbValue = sizeof(awcDeviceKey);
+    Status = RegQueryValue(hkey, pGraphicsDevice->szNtDeviceName, REG_SZ, awcDeviceKey, &cbValue);
+    ZwClose(hkey);
+    if (!NT_SUCCESS(Status))
+    {
+        ERR("Could not read '%S' registry value: 0x%08lx\n", Status);
+        return FALSE;
+    }
+
+    /* Replace Device0 by Video */
+    lastBkSlash = wcsrchr(awcDeviceKey, L'\\');
+    wcscpy(lastBkSlash + 1, L"Video");
```

Comment:
Do we have any guarantee that this wcscpy won't overrun awcDeviceKey[256]?

Reply:
This value (ending with `DeviceX`) is rewritten on each boot by videoprt.sys, and only SYSTEM and Administrators can change it.

Reply:
I would still prefer we didn't _ever_ use  wcscpy. `RtlStringCbCopyW` or similar is much superior.